File: LocalizationTester/MockCustomerContext.m
//
//  MockCustomerContext.m
//  LocalizationTester
//
//  Created by Cameron Sabol on 12/14/18.
//  Copyright © 2018 Stripe, Inc. All rights reserved.
//

#import "MockCustomerContext.h"

#pragma mark -  MockCustomer

@interface MockCustomer: NSObject
@property (nonatomic) NSMutableArray<STPPaymentMethod *> *mockPaymentMethods;
@property (nonatomic) STPPaymentMethod *mockDefaultPaymentMethod;
@property (nonatomic) STPAddress *mockShippingAddress;

@end

@implementation MockCustomer

- (instancetype)init {
    self = [super init];
    if (self) {
        _mockPaymentMethods = [NSMutableArray array];
        /**
         //     Preload the mock customer with saved cards.
         //     last4 values are from test cards: https://stripe.com/docs/testing#cards
         //     Not using the "4242" and "4444" numbers, since those are the easiest
         //     to remember and fill.
         //     */
        NSDictionary *visa = @{ @"card" : @{
                                    @"exp_month": @10,
                                    @"exp_year": @2020,
                                    @"last4": @"1881",
                                    @"brand": @"visa",
                                },
                                @"id": @"preloaded_visa",
                                @"type": @"card",
                                };

        STPPaymentMethod *visaCard = [STPPaymentMethod decodedObjectFromAPIResponse:visa];
        if (visaCard) {
            [_mockPaymentMethods addObject:visaCard];
        }

        NSDictionary *masterCard = @{ @"card" : @{
                                        @"exp_month": @10,
                                        @"exp_year": @2020,
                                        @"last4": @"8210",
                                        @"brand": @"mastercard",
                                        },
                                @"id": @"preloaded_mastercard",
                                @"type": @"card",
                                };
        STPPaymentMethod *masterCardCard = [STPPaymentMethod decodedObjectFromAPIResponse:masterCard];
        if (masterCardCard) {
            [_mockPaymentMethods addObject:masterCardCard];
        }

        NSDictionary *amex = @{ @"card" : @{
                                              @"exp_month": @10,
                                              @"exp_year": @2020,
                                              @"last4": @"0005",
                                              @"brand": @"amex",
                                              },
                                      @"id": @"preloaded_amex",
                                      @"type": @"card",
                                      };
        STPPaymentMethod *amexCard = [STPPaymentMethod decodedObjectFromAPIResponse:amex];
        if (amexCard) {
            [_mockPaymentMethods addObject:amexCard];
        }
    }

    return self;
}

- (NSArray<STPPaymentMethod *> *)paymentMethods {
    return self.mockPaymentMethods;
}

- (STPPaymentMethod *)defaultSource {
    return self.mockDefaultPaymentMethod;
}

- (STPAddress *)shippingAddress {
    return self.mockShippingAddress;
}

@end

#pragma mark - MockCustomerContext

@implementation MockCustomerContext
{
    MockCustomer *_mockCustomer;
}

- (instancetype)init {
    
    self = [super init];
    if (self) {
        _mockCustomer = [[MockCustomer alloc] init];
    }

    return self;
}

- (void)retrieveCustomer:(void (^ _Nullable)(STPCustomer * _Nullable, NSError * _Nullable))completion {
    if (!self.neverRetrieveCustomer) {
        if (completion) {
            completion((STPCustomer *)_mockCustomer, nil);
        }
    }
}

- (void)updateCustomerWithShippingAddress:(STPAddress * _Nonnull)shipping completion:(void (^ _Nullable)(NSError * _Nullable))completion {
    _mockCustomer.mockShippingAddress = shipping;
    if (completion) {
        completion(nil);
    }
}

- (void)listPaymentMethodsForCustomerWithCompletion:(void (^ _Nullable)(NSArray<STPPaymentMethod *> * _Nullable, NSError * _Nullable))completion {
    if (!self.neverRetrieveCustomer) {
        completion(_mockCustomer.mockPaymentMethods, nil);
    }
}

- (void)attachPaymentMethodToCustomer:(STPPaymentMethod * _Nonnull)paymentMethod completion:(void (^ _Nullable)(NSError * _Nullable))completion {
    [_mockCustomer.mockPaymentMethods addObject:paymentMethod];
    if (completion) {
        completion(nil);
    }
}

- (void)detachPaymentMethodFromCustomer:(STPPaymentMethod * _Nonnull)paymentMethod completion:(void (^ _Nullable)(NSError * _Nullable))completion {
    NSUInteger index = [_mockCustomer.mockPaymentMethods indexOfObjectPassingTest:^BOOL(STPPaymentMethod * _Nonnull obj, __unused NSUInteger idx, BOOL * _Nonnull stop) {
        if (obj.stripeId == paymentMethod.stripeId) {
            *stop = YES;
            return YES;
        }
        return NO;
    }];
    if (index != NSNotFound) {
        [_mockCustomer.mockPaymentMethods removeObjectAtIndex:index];
    }
    if (completion) {
        completion(nil);
    }
}

@end

File: Example/UI Examples/UI Examples/Source/MockCustomerContext.swift
//
//  MockCustomerContext.swift
//  UI Examples
//
//  Created by Ben Guo on 7/18/17.
//  Copyright © 2017 Stripe. All rights reserved.
//

import Foundation

@testable import Stripe
@_spi(STP) import StripeCore
@testable @_spi(STP) import StripePayments
@testable @_spi(STP) import StripePaymentsUI

class MockCustomer: STPCustomer {
    var mockPaymentMethods: [STPPaymentMethod] = []
    var mockDefaultPaymentMethod: STPPaymentMethod?
    var mockShippingAddress: STPAddress?

    init() {
        super.init(
            stripeID: "", defaultSource: nil, sources: [], shippingAddress: nil, email: nil,
            allResponseFields: [:])
        /**
         Preload the mock customer with saved cards.
         last4 values are from test cards: https://stripe.com/docs/testing#cards
         Not using the "4242" and "4444" numbers, since those are the easiest
         to remember and fill.
        */
        let visa =
            [
                "card": [
                    "id": "preloaded_visa",
                    "exp_month": "10",
                    "exp_year": "2020",
                    "last4": "1881",
                    "brand": "visa",
                ],
                "type": "card",
                "id": "preloaded_visa",
            ] as [String: Any]
        if let card = STPPaymentMethod.decodedObject(fromAPIResponse: visa) {
            mockPaymentMethods.append(card)
        }
        let masterCard =
            [
                "card": [
                    "id": "preloaded_mastercard",
                    "exp_month": "10",
                    "exp_year": "2020",
                    "last4": "8210",
                    "brand": "mastercard",
                ],
                "type": "card",
                "id": "preloaded_mastercard",
            ] as [String: Any]
        if let card = STPPaymentMethod.decodedObject(fromAPIResponse: masterCard) {
            mockPaymentMethods.append(card)
        }
        let amex =
            [
                "card": [
                    "id": "preloaded_amex",
                    "exp_month": "10",
                    "exp_year": "2020",
                    "last4": "0005",
                    "brand": "amex",
                ],
                "type": "card",
                "id": "preloaded_amex",
            ] as [String: Any]
        if let card = STPPaymentMethod.decodedObject(fromAPIResponse: amex) {
            mockPaymentMethods.append(card)
        }
    }

    var paymentMethods: [STPPaymentMethod] {
        get {
            return mockPaymentMethods
        }
        set {
            mockPaymentMethods = newValue
        }
    }

    var defaultPaymentMethod: STPPaymentMethod? {
        get {
            return mockDefaultPaymentMethod
        }
        set {
            mockDefaultPaymentMethod = newValue
        }
    }

    override var shippingAddress: STPAddress? {
        get {
            return mockShippingAddress
        }
        set {
            mockShippingAddress = newValue
        }
    }
}

class MockKeyProvider: NSObject, STPCustomerEphemeralKeyProvider {
    func createCustomerKey(
        withAPIVersion apiVersion: String, completion: @escaping STPJSONResponseCompletionBlock
    ) {
        completion(nil, NSError.stp_ephemeralKeyDecodingError())
    }
}
